### ISSUE TITLE
Access HttpServeletRequest in ResourceModel

### ISSUE THREAD

**Issue by cduque89** [unic/neba#164]
Hi all.

I am using the NEBA ResourceModel annotation in one of my AEM components.

This component loads the products from an AEM commerce collection and draws a carousel showing all the content.

The component can be dragged and dropped in a page by an author or can be placed in an html page programmatically using:

```html
<sly data-sly-resource="${'mobilePlansContainer' @ resourceType='gowingmyself/components/content/business/vf_mobilePlanSummary', collectionData='/etc/commerce/collections/Bill Pay Plans'}"></sly>
```

When we are dragging and dropping the component the ResourceModel works perfectly:

```java
@ResourceModel(types = {"gowingmyself/components/content/business/vf_mobilePlanSummary"})
public class OfferListModel {

    @Autowired
    private ProductCatalogAgent productCatalogAgent;

    @Autowired
    private ModelMapper mapper;

    @This
    private Resource resource;

    @Path("collectionData")
    private String path;
    
    private List<OfferModel> offers;

    @PostMapping
    public void postMapping() {
        offers = getOfferList();
    }

    private List<OfferModel> getOfferList() {

        List<OfferModel> offerListModel = new ArrayList<>();

        if (isValidPath()) {

            try {
                List<OmnichannelProductImpl> productCollection = productCatalogAgent.getProductCollection(resource, this.path);

                for (OmnichannelProductImpl product: productCollection) {
                    OfferModel offerModel = mapper.productImplToOfferModel(product);
                    setModelProperties(product, offerModel);

                    offerListModel.add(offerModel);
                }

            } catch (CommerceException e) {
                e.printStackTrace();
            }
        }

        return offerListModel;
    }

    private boolean isValidPath() {
        return !Strings.isNullOrEmpty(this.path);
    }

    private void setModelProperties(OmnichannelProductImpl product, OfferModel offerModel) {
        mapper.updateOfferModelFromProduct(product, offerModel);
    }
}
```

We can access the `@Path("collectionData") private String path;`  and it is filled with the correct value that was set by the author in a Dialog.

But when we are adding the component programmatically the @PostMapping code runs, the `@Path("collectionData") private String path;` is not being filled (which we understand), but we need to access e some way.

We were expecting to obtain it from the HttpServletRequest but we cannot find a way to access to it inside the @ResourceModel.

Do you have any solution for this? Or are we doing this in the wrong way?

Thanks in advance for your help.

Regards,
Carlos Duque

**Comment by olaf-otto** [comment]
Hi @cduque89, thanks for reaching out! That looks like a very interesting use case.

Indeed, NEBA does not provide a default way to access the HTTP request / response in models. The reason being is that a model is associated with a resource, which is agnostic of the servlet context. This is intended to further a much better separation of concern, as we observed that mixing request state and models created controller / model mashups in the past, which led a significant decrease of separation of concerns, tight coupling of model state with complex request state and so forth.

That being said, I'm sure there is an elegant solution to your requirement! Could you explain what you mean by "adding the component programmatically", perhaps providing a code sample?

Thanks & regards,
Olaf

**Comment by cduque89** [comment]
Hi @olaf-otto. 

Thanks for your answer.

I am referring to the first code snippet that I placed. The one using the data-sly-resource to place the component in the page. 

I think that I am probably falling in this situation:
"we observed that mixing request state and models created controller / model mashups in the past".

I am open for a more "elegant" solution as you said :).

Regards,
Carlos

**Comment by olaf-otto** [comment]
Hi @cduque89,

Ahh, I see - you are providing a default value for collectionData in the inclusion to be used if no value for the property exists in the resource, correct?

I am travelling at the moment but will get back to you with some ideas as soon as possible.

Kind regards,
Olaf

**Comment by cduque89** [comment]
Hi @olaf-otto.

Yes that's it. The authors can drag and drop the component in any place and choose a product collection. But at the same time we have some pages where we want to show some specific collections without the authoring input. In those cases we want to programmatically place the component in the page and give a default collectionData.

Imagine the scenarios:

1. The author drags and drops the component in the home page as chooses a product collection that contains phones from Apple. But tomorrow they can go there and change it to a collection with phones from Samsung. Or create a different collection and mix all the phones. This is the scenario that works perfectly :).

2. The second scenario, which is the one that is not working for us, is having a page that will always show the same collection in the component and the author cannot change it. In this case we need to programmatically place the component in the specific page and pass a parameter with the collection to render. As I said in the first comment we can place a breakpoint in the @PostMapping method and stop there, but we are not finding a way to have access to the collectionData parameter that we are passing to the component. 

I think this description gives you a better understanding of what we want to achieve.

Thanks again for your support.

Regards,
Carlos

**Comment by olaf-otto** [comment]
Hi @cduque89,

From my point of view, there is no need for a resource model for what you are trying to do, as the main functionality modeled here is "resolve product collection by (arbitrary) path". The only relation between the specific resource and that use case is that the resource _may_ provide a collection's path. I would thus recommend to instead publish an OSGi service providing a service method like 

`List<OfferModel> getOffers(String path, ...)`

Then, you can simply use that service in your resource's view, for instance using a JS via the use API, like so:

```
// myScript.js
use(function () {
    // Java packages are accessed from JS using Packages.<FQCN>
    var service = sling.getService(Packages.my.package.OfferListService);
    return service.getOffers(this.collectionData, resource);
});
```

collectionData could then be passed to the script from a resource property of a default value, like so:

`<... data-sly-use.offers='${"myScript.js" @ collectionData="/etc/commerce/collections/Bill Pay Plans"}'...>`

However, if you want to go the other way and better support further commerce API use cases in models, I would recommend using custom annotations via a [custom field mapper](http://neba.io/apidocs/neba-api/io/neba/api/resourcemodels/AnnotatedFieldMapper.html). This custom mapper could access the HttpRequest via Spring's static RequestContextHolder and abstract the loading and provisioning of, for instance, OfferModel instances previously implemented in the model itself. The resulting model might look like so:

```
@ResourceModel(types = {"gowingmyself/components/content/business/vf_mobilePlanSummary"})
public class OfferListModel {
    @Path("collectionData")
    @Offers
    private List<OfferModel> offers;
}
```

With `@Offers ` being a custom annotation supported by a specific AnnotatedFieldMapper service implementation. 

However, I would advise going with the OSGi service approach.

One more thing: Why does your model extend DefineObjectsTag?

Kind regards,
Olaf

**Comment by cduque89** [comment]
Hi.

The DefineObjectsTag was just something we placed there to view the code of the class. I will change that in the first comment (it is not there anymore).

We are going the analyse your suggestions more carefully and try to understand which one suits us better.

Will get back to you with more questions or with the solutions we implemented.

Thanks a lot for your support.

Regards,
Carlos

**Comment by cduque89** [comment]
Hi again @olaf-otto.

We followed another approach. We used the WCMUsePojo class.

It as a similar behaviour as the @ResourceModel but we can access the HttpServletRequest from there.

We tried your suggestion with the OSGi service component but we had some problems accessing the ProductCatalogAgent service that was @Autowired in the first code example of the @ResourceModel that I sent to you. I think this would be a more clean approach as you said before.

Thanks a lot for your help.

Regards,
Carlos
